### PR TITLE
COS-228 Added 'whereCondition' field to 'remove documents' change in order to filter documents via sql++ clause

### PR DESCRIPTION
**Pull request by tigrankhojoyan**
…order to filter documents via sql++ clause

### CODE REVIEW

File: liquibase-couchbase/src/main/java/liquibase/ext/couchbase/statement/RemoveDocumentsQueryStatement.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    public RemoveDocumentsQueryStatement(Keyspace keyspace, List<Id> ids, String whereCondition) {
+        super(keyspace, ids);
+        this.whereCondition = whereCondition;
+    }
+
+    @Override
+    public void doInTransaction(TransactionAttemptContext transaction, ClusterOperator clusterOperator) {
+        List<String> filteredIds = clusterOperator.retrieveDocumentIdsByWhereClause(getKeyspace(), whereCondition);
+        getIds().addAll(filteredIds.stream().map(Id::new).collect(Collectors.toList()));
```
Comment (DaimonCool) [inline liquibase-couchbase/src/main/java/liquibase/ext/couchbase/statement/RemoveDocumentsQueryStatement.java:29]: Maybe we should convert all this to set to remove duplicate ids ?
Imagine that we pass ids and whereClause in changeset. And we already have id "key1" in ids list. And where query also found document with id "key1" and added into list. We will have 2 duplicate ids in the end.

File: liquibase-couchbase/src/test/java/integration/statement/RemoveDocumentsQueryStatementIT.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static common.constants.TestConstants.INDEX;
+import static common.matchers.CouchbaseCollectionAssert.assertThat;
+import static liquibase.ext.couchbase.types.Keyspace.keyspace;
+
+class RemoveDocumentsQueryStatementIT extends TransactionStatementTest {
+
+    private static final String DOC_FIELD_NAME = "field";
+    private static final String DOC_FIELD_Value = "val";
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/integration/statement/RemoveDocumentsQueryStatementIT.java:29]: DOC_FIELD_Value -> DOC_FIELD_VALUE

File: liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/RemoveDocumentsChangeTest.java [hunk 3]
```diff
@@ -65,5 +66,16 @@ void Should_contains_exactly_one_document_id() {
...
     }
+
+    @Test
+    void Should_contain_where_clause() {
+        DatabaseChangeLog changeLog = changeLogProvider.load(REMOVE_BY_QUERY_TEST_XML);
+        ChangeSet changeSet = firstOf(changeLog.getChangeSets());
+        List<Change> changes = changeSet.getChanges();
+        assertThat(changes).hasSize(1);
+        assertThat(((RemoveDocumentsChange) changes.get(0)).getIds()).hasSize(0);
+        assertThat(((RemoveDocumentsChange) changes.get(0)).getWhereCondition()).isEqualTo("test=\"test\"");
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/liquibase/ext/couchbase/change/RemoveDocumentsChangeTest.java:77]: extract ((RemoveDocumentsChange) changes.get(0)) to seperate variable

File: liquibase-couchbase/src/test/java/system/change/RemoveDocumentsSystemTest.java [hunk 4]
```diff
@@ -26,6 +31,9 @@
 
 public class RemoveDocumentsSystemTest extends LiquibaseSystemTest {
 
+    private static final String DOC_FIELD_NAME = "field1";
+    private static final String DOC_FIELD_Value = "val1";
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/system/change/RemoveDocumentsSystemTest.java:35]: DOC_FIELD_Value -> DOC_FIELD_VALUE

File: liquibase-couchbase/src/test/java/system/change/RemoveDocumentsSystemTest.java [hunk 5]
```diff
@@ -40,6 +48,18 @@ private void insertTestDocuments() {
...
+    private void insertDocsUsingSpecificFields() throws InterruptedException {
+        collectionOperator.createPrimaryIndex(CreatePrimaryQueryIndexOptions
+                .createPrimaryQueryIndexOptions()
+                .indexName(INDEX));
+        TimeUnit.SECONDS.sleep(3L);
+        Document document1 = collectionOperator.generateTestDocByBody(JsonObject.create().put(DOC_FIELD_NAME, DOC_FIELD_Value));
+        Document document2 = collectionOperator.generateTestDocByBody(JsonObject.create().put(DOC_FIELD_NAME, DOC_FIELD_Value));
+        collectionOperator.insertDocs(document1, document2);
+        ids = Lists.newArrayList(new Id(document1.getId()), new Id(document2.getId()));
+        TimeUnit.SECONDS.sleep(3L);
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/system/change/RemoveDocumentsSystemTest.java:60]: I think test will be passed without this sleep. Please check

File: liquibase-couchbase/src/test/java/integration/statement/RemoveDocumentsQueryStatementIT.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    private Keyspace keyspace;
+
+    @BeforeEach
+    @SneakyThrows
+    void setUp() {
+        keyspace = keyspace(bucketName, scopeName, collectionName);
+        collection = bucketOperator.getCollection(collectionName, scopeName);
+        collectionOperator.createPrimaryIndex(CreatePrimaryQueryIndexOptions
+                .createPrimaryQueryIndexOptions()
+                .indexName(INDEX));
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/integration/statement/RemoveDocumentsQueryStatementIT.java:45]: create primary index 1 time for all tests in this class. No point to create it and drop for each test

File: liquibase-couchbase/src/test/java/system/change/RemoveDocumentsSystemTest.java [hunk 7]
```diff
@@ -48,13 +68,28 @@ void cleanUpd() {
         if (collectionOperator.docExists(docId2)) {
             collectionOperator.removeDoc(doc2);
         }
+        if (collectionOperator.collectionIndexExists(INDEX)) {
+            collectionOperator.dropIndex(INDEX);
+        }
```
Comment (DaimonCool) [inline liquibase-couchbase/src/test/java/system/change/RemoveDocumentsSystemTest.java:73]: We can drop index after all tests. No need to create and drop it every time

File: liquibase-couchbase/src/main/java/liquibase/ext/couchbase/statement/RemoveDocumentsQueryStatement.java (hunk 1 above)
Comment (tigrankhojoyan) [reply liquibase-couchbase/src/main/java/liquibase/ext/couchbase/statement/RemoveDocumentsQueryStatement.java:30]: Agree, changed IDs container from List to Set

File: liquibase-couchbase/src/test/java/system/change/RemoveDocumentsSystemTest.java (hunk 7 above)
Comment (tigrankhojoyan) [reply liquibase-couchbase/src/test/java/system/change/RemoveDocumentsSystemTest.java]: Done, thanks!

File: liquibase-couchbase/src/test/java/integration/statement/RemoveDocumentsQueryStatementIT.java (hunk 6 above)
Comment (DaimonCool) [reply liquibase-couchbase/src/test/java/integration/statement/RemoveDocumentsQueryStatementIT.java:45]: Not fixed yet.

Comment (tigrankhojoyan) [reply liquibase-couchbase/src/test/java/integration/statement/RemoveDocumentsQueryStatementIT.java:53]: Sorry, missed this part. Fixed now)

**Review by CountRick — APPROVED** [review]

**Review by DaimonCool — APPROVED** [review]

**Pull request merged**